### DISCUSSION TITLE
Feature Request: secondary, package manager-specific extension directory (e.g. for conda)

### DISCUSSION THREAD

**Discussion by xhochy** [duckdb/duckdb#15352] [Ideas]
I would like to provide DuckDB extensions as conda-forge packages instead of using the official distribution method. This has multiple advantages for us:

1. We can use the existing software proxying channels we use for the rest of our software dependencies into "internet-disconnected" environments.
2. We can list them in our [`pixi.toml`](https://pixi.sh/latest/) 
3. We can built multiple versions of them as [archspec-enabled packages](https://github.com/conda-forge/conda-forge.github.io/issues/1261) so that the binaries match the best instruction set the user has available.
4. We could (theoretically) unbundle some of the statically linked third-party dependencies as we have full control over all native libraries in a conda setting.

From what I have understood in the current "packaging" setup of the extensions is that they are user-wide installed by default and also requires a cryptographic signature. Both of them are not available in a conda-forge based setup. Thus, my proposal would be to have an optional flag in duckdb to provide:

1. A secondary extension directory which is also used for extension lookups but is not the one where extensions are installed into.
2. Allow unsigned extensions to be loaded from this directory only, even if you don't run using the `-unsigned` flag.

_I'm happy to implement all of this, I'm mostly looking for early feedback on that path I'm taking_

**Comment by teaguesterling** [comment]
I've thought about this as well while packaging duckdb for spack. Delivering a consistent duckdb install with extensions to multiple users (especially in an air-gapped environment) is more difficult than it should be. I looked at some options for implementing it and had some ideas. The user-specific directory is hard-coded as a fallback, but it shouldn't be too hard to add some additional checks.

I've been thinking about this as a compile time option to make extensions install-wide, but it seems you're thinking of it as more of a runtime option. Given that we both are interested in this but also have slightly different use cases, maybe we can work together on this to make sure it's flexible enough. Happy to collaborate on implementing!

**Comment by hannes** [comment]
We're not able to allow unsigned extension loads without the `-unsigned` flag for security reasons.  As for the extension directory, it is already configurable, why is a second one needed?

**Reply by teaguesterling** [reply]
Signed or unsigned wouldn't matter for my use case. The primary need is to be able to maintain a consistent set of extensions for a single install used by multiple users in an air-gapped environment. Having a second default extensions directory that's not username specific would allow an admin to pre-install all the necessary (signed or unsigned) extensions and have things just work without requiring users to run additional idiosyncratic installation steps each time.

**Reply by xhochy** [reply]
For my use case, unsigned would actually matter as we want to build the extensions as conda packages and don't have the facilities to sign them in the process. I envision this to be split, though: For the primary extension directory (where user installs go), you would require signed extensions, and for the secondary directory, which is only filled with extensions installed by the package manager, you would be OK with not signing the extensions.

**Comment by teaguesterling** [comment]
@xhochy for the first point, I'd be happy to work with you adding in a compile time option to check a second install-wide directory for extensions. For the second point, I think @hannes point stands that you may not want to allow that as part of the compilation. I'll bet if you poked around you *could* find how to patch the source during your compilation to allow loading extensions signed from a 3rd party, but that could be risky.

**Reply by xhochy** [reply]
Sorry for dropping the ball. Christmas and so on got in between and happy to pick this up again. For me the allowing unsigned extensions would be crucial to get this to work smoothly. I know how I could circumvent this, but I'd prefer to work with upstream here instead of building an inoffical/hidden feature for conda here.

**Reply by teaguesterling** [reply]
No worries. Maybe we should split this into two steps. One for the install-wide extension directory and another for a compile time option to disable extension signing checks.

**Reply by xhochy** [reply]
Finally got the time to write a first draft: https://github.com/duckdb/duckdb/pull/17499 

This works nicely in our conda setup, but tests are missing and I don't understand why the constant string has two trailing nulls on Linux.

**Comment by nicornk** [comment]
Having duckdb and its extensions available through conda-forge would very much ease the adoption of duckdb in enterprise environments which are often air-gapped.
